In [1]:
# Data Wrangling
import pandas as pd
from pandas import Series, DataFrame
import numpy as np

# Visualization
import matplotlib.pylab as plt
from matplotlib import font_manager, rc
import seaborn as sns
%matplotlib inline

# EDA
# import klib

# Preprocessing & Feature Engineering
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.impute import SimpleImputer 
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import PowerTransformer
from sklearn.feature_selection import SelectPercentile
from sklearn import base
from sklearn.model_selection import KFold
from sklearn.model_selection import StratifiedShuffleSplit
from sklearn.model_selection import StratifiedKFold
from sklearn.experimental import enable_iterative_imputer  # still experimental 
from sklearn.impute import IterativeImputer
from sklearn.feature_selection import RFE


# Hyperparameter Optimization
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RandomizedSearchCV

# Modeling
from sklearn.dummy import DummyClassifier
from sklearn.svm import SVR
from lightgbm import LGBMClassifier
from sklearn.linear_model import Ridge, Lasso, ElasticNet
from sklearn.linear_model import BayesianRidge

# Evaluation
from sklearn.metrics import roc_auc_score
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score
from sklearn.metrics import mean_squared_error
from sklearn.metrics import log_loss

# Utility
import os
import time
import random
import sys, warnings
if not sys.warnoptions: warnings.simplefilter("ignore")
from IPython.display import Image
# import pickle
from tqdm import tqdm
import platform
from itertools import combinations
from scipy.stats.mstats import gmean
from tensorflow import keras

# from bayes_opt import BayesianOptimization

In [2]:
X_new = pd.read_csv(os.path.abspath("../input")+"/hwang_selected_features_train.csv" , encoding = 'utf-8')
X_te_new = pd.read_csv(os.path.abspath("../input")+"/hwang_selected_features_test.csv" , encoding = 'utf-8')

In [50]:
target = pd.read_csv(os.path.abspath("../input")+'/y_train.csv' , encoding = 'cp949').group

In [16]:
# dictionary에 있는 target값을 모두 추출
target_list = []

for result in lgbBO.res:
    target = result['target']
    target_list.append(target)
print(target_list)
# 가장 큰 target 값을 가지는 순번(index)를 추출
print('maximum target index:', np.argmin(np.array(target_list)))

[1.6060982858125845, 1.4556068675499798, 1.4655825658321555, 1.4688808847928827, 1.4591822611433771, 1.4669657412254045, 1.452348853289662, 1.4645161000335565, 1.4626696571767683, 1.4799900664906072, 1.5443871109560476, 1.4671639551435856, 1.5274123757440137, 1.4675921068085074, 1.6950039702477975, 1.4933062656887393, 1.638815821205718, 1.4627763950780488, 1.4680463155627095, 1.4662920486477267, 1.4668566526359093, 1.4767312387673706, 1.5428294883105653, 1.5758729620943188, 1.4624938648355332, 1.4781610996959282, 1.4854362789477158, 1.468151754343313, 1.459436314821971, 1.4713478269480162, 1.4730684078905438, 1.4642859113913247, 1.4891334602953101, 1.4998583551808504, 1.461139181123826, 1.4627578493200197, 1.4619787961882376, 1.4690911586541557, 1.467681102742748, 1.4632863093789008, 1.4727158703805407, 1.4774825834921883, 1.4877343294300907, 1.5913676784636825, 1.4607328430612407, 1.4634965955517054, 1.4699443732746134, 1.470698345132294, 1.4753763253485417, 1.4921084861715843, 1.4671

In [17]:
# 가장 큰 target값을 가지는 index값을 기준으로 res에서 해당 parameter 추출. 
max_dict = lgbBO.res[np.argmin(np.array(target_list))]
print(max_dict)

{'target': 1.452348853289662, 'params': {'colsample_bytree': 0.5489297647691318, 'learning_rate': 0.033947270113616715, 'max_bin': 64.96058512468377, 'max_depth': 13.838623612170684, 'min_child_samples': 65.50145788216062, 'min_child_weight': 37.87286066499082, 'n_estimators': 355.00257308928775, 'num_leaves': 48.02022146460943, 'reg_alpha': 21.197890290451262, 'reg_lambda': 9.521528307448555, 'subsample': 0.5990212685265623}}


# *OOF 스태킹*

In [30]:
from sklearn.preprocessing import LabelEncoder

In [31]:
encoder = LabelEncoder()

In [32]:
y = encoder.fit_transform(target)

In [33]:
X_new.shape, X_te_new.shape

((21587, 875), (14380, 875))

In [34]:
lgb_oof_ver2 = np.zeros((X_new.shape[0], 19))
lgb_oof_ver2

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [35]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from tqdm import tqdm
from time import time
import datetime
import gc

# FE
from scipy.signal import find_peaks, peak_widths, peak_prominences

# Model
from sklearn.model_selection import StratifiedKFold
from sklearn.preprocessing import LabelEncoder

import lightgbm as lgb
import xgboost as xgb

# Ensemble
from scipy.optimize import minimize
from sklearn.metrics import log_loss

{'target': 1.452348853289662, 'params': {'colsample_bytree': 0.5489297647691318, 'learning_rate': 0.033947270113616715, 'max_bin': 64.96058512468377, 'max_depth': 13.838623612170684, 'min_child_samples': 65.50145788216062, 'min_child_weight': 37.87286066499082, 'n_estimators': 355.00257308928775, 'num_leaves': 48.02022146460943, 'reg_alpha': 21.197890290451262, 'reg_lambda': 9.521528307448555, 'subsample': 0.5990212685265623}}

In [36]:
lgbm_params= {'num_leaves': 35,
        'min_child_samples': 182,
        'min_child_weight': 26,
        'max_depth': 10,
        'learning_rate': 0.018668206381448577,
        'n_estimators': 361,
        'reg_lambda': 0.6690395135477213,
        'reg_alpha': 0.7723190057522233,
        'subsample': 0.8855706663296421}

In [37]:
clf = LGBMClassifier( **lgbm_params,
                objective='multiclass',
                metric='multi_logloss',
                nthread=4,
                random_state = 0
                )

In [38]:
n_splits = 10
mlogloss = []
lgb_oof_ver2 = np.zeros((X_new.shape[0], 8))
lgb_pred_ver2 = np.zeros((X_te_new.shape[0], 8))

for X, X_test in [(X_new,X_te_new)]:
    X= X.reset_index(drop=True)
    for seed in [0,1000]:
        kfold = StratifiedKFold(n_splits=n_splits, random_state= seed, shuffle=True)
        for fold, (trn_idx, val_idx) in enumerate(kfold.split(X, y)):
            X_train, X_valid = X.iloc[trn_idx], X.iloc[val_idx]
            y_train, y_valid = y[trn_idx], y[val_idx]

            clf.fit(X_train, y_train, eval_set=[(X_train, y_train), (X_valid, y_valid)], eval_metric= 'logloss', verbose= 200, 
                early_stopping_rounds= 200)
            # Predict
            lgb_pred_ver2 += clf.predict_proba(X_test) / (n_splits * 4)
            lgb_oof_ver2[val_idx] += clf.predict_proba(X_valid) / 4
            print('*'* 85)
            print('Training has finished.')
        print('lgb ver2 logloss= ', log_loss(y, lgb_oof_ver2))

[LightGBM] [Warning] num_threads is set with n_jobs=-1, nthread=4 will be ignored. Current value: num_threads=-1
[200]	training's multi_logloss: 1.01914	valid_1's multi_logloss: 1.49866
*************************************************************************************
Training has finished.
[LightGBM] [Warning] num_threads is set with n_jobs=-1, nthread=4 will be ignored. Current value: num_threads=-1
[200]	training's multi_logloss: 1.01647	valid_1's multi_logloss: 1.50977
*************************************************************************************
Training has finished.
[LightGBM] [Warning] num_threads is set with n_jobs=-1, nthread=4 will be ignored. Current value: num_threads=-1
[200]	training's multi_logloss: 1.0169	valid_1's multi_logloss: 1.50876
*************************************************************************************
Training has finished.
[LightGBM] [Warning] num_threads is set with n_jobs=-1, nthread=4 will be ignored. Current value: num_threads=-1
[20

In [39]:
all_oof = np.column_stack([lgb_oof_ver2])
all_test = np.column_stack([lgb_pred_ver2])

In [40]:
pd.DataFrame(all_test)

0         1         2         3         4         5         6  \
0      0.047789  0.055238  0.039651  0.058313  0.022823  0.215682  0.024095   
1      0.038717  0.121665  0.158328  0.012561  0.002407  0.037217  0.110800   
2      0.390810  0.075028  0.008975  0.005140  0.007648  0.006072  0.002036   
3      0.269639  0.043306  0.023227  0.026753  0.010771  0.110532  0.006263   
4      0.410867  0.045430  0.010706  0.003513  0.016392  0.008330  0.002798   
...         ...       ...       ...       ...       ...       ...       ...   
14375  0.091040  0.210153  0.010941  0.008008  0.008774  0.127177  0.034096   
14376  0.188225  0.123701  0.047963  0.036643  0.017691  0.036518  0.032995   
14377  0.178846  0.097681  0.061746  0.046816  0.032216  0.044802  0.020240   
14378  0.080011  0.220279  0.024645  0.006511  0.014394  0.130358  0.020012   
14379  0.342658  0.046557  0.017183  0.004789  0.034398  0.039882  0.007928   

              7  
0      0.036409  
1      0.018306  
2      0.004292  
3      0.009509  
4      0.001964  
...         ...  
14375  0.009811  
14376  0.016264  
14377  0.017652  
14378  0.003790  
14379  0.006605  

[14380 rows x 8 columns]

In [41]:
X_train.shape

(19429, 875)

In [42]:
all_test.shape

(14380, 8)

In [43]:
mlogloss = []
n_splits = 10

stk_oof_pred = np.zeros((all_oof.shape[0], 8))
stk_test_pred = np.zeros((all_test.shape[0], 8))

kfold = StratifiedKFold(n_splits=n_splits, random_state=1000, shuffle=True)
for fold, (trn_idx, val_idx) in enumerate(kfold.split(all_oof, y)):
    X_train, X_valid = all_oof[trn_idx], all_oof[val_idx]
    y_train, y_valid = y[trn_idx], y[val_idx]
            
    clf.fit(X_train, y_train, eval_set=[(X_train, y_train), (X_valid, y_valid)], eval_metric= 'logloss', verbose= 200, 
                early_stopping_rounds= 200)


    stk_test_pred += clf.predict_proba(all_test) / n_splits
    stk_oof_pred[val_idx] = clf.predict_proba(X_valid)
    
print('mean logloss= ',np.mean(mlogloss))

[LightGBM] [Warning] num_threads is set with n_jobs=-1, nthread=4 will be ignored. Current value: num_threads=-1
[200]	training's multi_logloss: 1.33502	valid_1's multi_logloss: 1.48007
[LightGBM] [Warning] num_threads is set with n_jobs=-1, nthread=4 will be ignored. Current value: num_threads=-1
[200]	training's multi_logloss: 1.33544	valid_1's multi_logloss: 1.47643
[LightGBM] [Warning] num_threads is set with n_jobs=-1, nthread=4 will be ignored. Current value: num_threads=-1
[200]	training's multi_logloss: 1.33061	valid_1's multi_logloss: 1.51607
[LightGBM] [Warning] num_threads is set with n_jobs=-1, nthread=4 will be ignored. Current value: num_threads=-1
[200]	training's multi_logloss: 1.33351	valid_1's multi_logloss: 1.46864
[LightGBM] [Warning] num_threads is set with n_jobs=-1, nthread=4 will be ignored. Current value: num_threads=-1
[200]	training's multi_logloss: 1.33213	valid_1's multi_logloss: 1.51352
[LightGBM] [Warning] num_threads is set with n_jobs=-1, nthread=4 will

In [44]:
stk_test_pred = pd.DataFrame(stk_test_pred)

In [45]:
stk_test_pred = stk_test_pred.rename(columns = {0:'F20',1:'F30',2:'F40',3:'F50',4:'M20',5:'M30',6:'M40',7:'M50'})

In [46]:
stk_test_pred

F20       F30       F40       F50       M20       M30       M40  \
0      0.110668  0.161862  0.061028  0.119777  0.047515  0.317853  0.082164   
1      0.046227  0.241898  0.405192  0.018253  0.006599  0.078895  0.174497   
2      0.808575  0.114521  0.015947  0.014242  0.017808  0.015102  0.005297   
3      0.486203  0.140725  0.045855  0.055188  0.061642  0.170928  0.017106   
4      0.798341  0.083580  0.023989  0.017055  0.035164  0.024162  0.005393   
...         ...       ...       ...       ...       ...       ...       ...   
14375  0.208378  0.376220  0.021342  0.011363  0.016453  0.308240  0.044508   
14376  0.383008  0.236703  0.152370  0.042342  0.036599  0.086330  0.034241   
14377  0.346019  0.265763  0.113701  0.084771  0.053381  0.080321  0.025236   
14378  0.184218  0.405808  0.037924  0.014140  0.024410  0.253406  0.073004   
14379  0.687258  0.072646  0.021399  0.011300  0.099569  0.084693  0.008812   

            M50  
0      0.099132  
1      0.028439  
2      0.008508  
3      0.022353  
4      0.012316  
...         ...  
14375  0.013495  
14376  0.028408  
14377  0.030810  
14378  0.007091  
14379  0.014322  

[14380 rows x 8 columns]

In [47]:
x_te = pd.read_csv('../input/X_test.csv', encoding='cp949')
tst_id = x_te['custid'].unique()

In [48]:
submission = pd.concat([pd.DataFrame({'ID':tst_id}),stk_test_pred],axis = 1)
submission.to_csv('hwang_lgbm_stk_oof10.csv',index = False,encoding = 'utf-8')

In [49]:
submission

ID       F20       F30       F40       F50       M20       M30  \
0      30001  0.110668  0.161862  0.061028  0.119777  0.047515  0.317853   
1      30002  0.046227  0.241898  0.405192  0.018253  0.006599  0.078895   
2      30003  0.808575  0.114521  0.015947  0.014242  0.017808  0.015102   
3      30005  0.486203  0.140725  0.045855  0.055188  0.061642  0.170928   
4      30007  0.798341  0.083580  0.023989  0.017055  0.035164  0.024162   
...      ...       ...       ...       ...       ...       ...       ...   
14375  49988  0.208378  0.376220  0.021342  0.011363  0.016453  0.308240   
14376  49990  0.383008  0.236703  0.152370  0.042342  0.036599  0.086330   
14377  49992  0.346019  0.265763  0.113701  0.084771  0.053381  0.080321   
14378  49993  0.184218  0.405808  0.037924  0.014140  0.024410  0.253406   
14379  49994  0.687258  0.072646  0.021399  0.011300  0.099569  0.084693   

            M40       M50  
0      0.082164  0.099132  
1      0.174497  0.028439  
2      0.005297  0.008508  
3      0.017106  0.022353  
4      0.005393  0.012316  
...         ...       ...  
14375  0.044508  0.013495  
14376  0.034241  0.028408  
14377  0.025236  0.030810  
14378  0.073004  0.007091  
14379  0.008812  0.014322  

[14380 rows x 9 columns]